datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java

```diff
@@ -337,12 +330,31 @@ private ZoneId getZone(DeserializationContext context)
     private String replaceZeroOffsetAsZIfNecessary(String text)
     {
         if (replaceZeroOffsetAsZ) {
-            return ISO8601_UTC_ZERO_OFFSET_SUFFIX_REGEX.matcher(text).replaceFirst("Z");
+            return replaceZeroOffsetAsZ(text);
         }
 
         return text;
     }
 
+    private static String replaceZeroOffsetAsZ(String text)
+    {
+        int plusIndex = text.lastIndexOf('+');
+        if (plusIndex < 0) {
+            return text;
+        }
+        int maybeOffsetIndex = plusIndex + 1;
+        int remaining = text.length() - maybeOffsetIndex;
+        if (remaining < 2 || remaining == 3 || remaining > 5) {
+            return text;
+        }
+
+        String maybeOffset = text.substring(maybeOffsetIndex);
+        if ("00".equals(maybeOffset) || "0000".equals(maybeOffset) || "00:00".equals(maybeOffset)) {
```

Comment:
Can we optimize this using `regionMatches`?
perhaps using something along the lines of
```java
switch (remaining) {
    case 2:
       return text.regionmatches..."00" ? text.substring(0, plusIndex) + 'Z' : text;
     case 4:
        ..etc
     case 5:
       ..etc
 }
 return text;
```

Reply:
Seems reasonable to avoid the substring allocation, updated to use `regionMatches` & `switch` and reran JMH and tested locally with JDKs 8, 11, 17.

```
Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
JDK 17.0.5, OpenJDK 64-Bit Server VM, 17.0.5+8-LTS
Benchmark                           Mode  Cnt    Score   Error  Units
InstantDeserializerBenchmark.index  avgt    5   20.003 ± 0.486  ns/op
InstantDeserializerBenchmark.regex  avgt    5  156.174 ± 7.654  ns/op
```

```
Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
JDK 11.0.16.1, OpenJDK 64-Bit Server VM, 11.0.16.1+9-LTS
Benchmark                           Mode  Cnt    Score   Error  Units
InstantDeserializerBenchmark.index  avgt    5   20.818 ± 1.267  ns/op
InstantDeserializerBenchmark.regex  avgt    5  192.248 ± 7.341  ns/op
```

```
Apple M1 Pro aarch64
JDK 17.0.6, OpenJDK 64-Bit Server VM, 17.0.6+10-LTS
Benchmark                           Mode  Cnt   Score   Error  Units
InstantDeserializerBenchmark.index  avgt    5  17.498 ± 0.677  ns/op
InstantDeserializerBenchmark.regex  avgt    5  58.096 ± 1.330  ns/op
```

```
Apple M1 Pro aarch64
Benchmark                           Mode  Cnt   Score   Error  Units
InstantDeserializerBenchmark.index  avgt    5  16.176 ± 7.842  ns/op
InstantDeserializerBenchmark.regex  avgt    5  63.201 ± 5.668  ns/op
```